// Marco Ermini - March 2021 for ASU CSE531 Course
// Do not leech!
// Implementation of a banking's branches/customers RPC synchronisation using GRPC and Python
// Built with python 3.8 with GRPC and GRPC-tools libraries; may work with other Python versions


syntax = "proto3";

package app;

service Banking {
	rpc MsgDelivery(MsgDeliveryRequest) returns (MsgDeliveryResponse) {}
//	rpc GetEvents(EventsRequest) returns (EventsResponse) {}
}

enum Operation {
	QUERY 			= 0;		// Query amount of $$
	DEPOSIT			= 1;		// Deposit amounf of $$
	WITHDRAW		= 2;		// Withdraw amount of $$
}

enum ReturnCode {
	SUCCESS 		= 0;		// Query, Deposit, or Withdraw succeeded
	FAILURE			= 1;		// Query, Deposit, or Withdraw failed
	ERROR			= 2;		// Error in the operation
}

enum SourceType {
	CUSTOMER 		= 0;		// Query, Deposit, or Withdraw succeeded
	BRANCH			= 1;		// Query, Deposit, or Withdraw failed
}

message MsgDeliveryRequest  {
	int64 REQ_ID	= 1;		// Branch or Client's Request ID
	Operation OP	= 2;		// Operation requested (query, deposit, withdraw)
	int64 Amount 	= 3;		// Amount involved in the operation
	int64 S_TYPE	= 4;		// Source Type (Customer or Branch)
	int64 S_ID		= 5;		// Source ID
	int64 D_ID		= 6;		// Branch Destination ID
	int64 Clock		= 7;		// Tick clock in the case of logical clock use
}

message MsgDeliveryResponse {
	int64 REQ_ID	= 1;		// Branch or Client's Request ID
	ReturnCode RC	= 2;		// Return Code from the operation
	int64 Amount	= 3;		// Amound of $$ involved
	int64 Clock		= 4;		// Tick clock in the case of logical clock use
}

// message EventsRequest {}

// message EventsResponse {
// 	repeated Event events = 1;
// }

// message Event {
// 	int64 ID = 1;
// 	string OP = 2;
// 	int64 Clock = 3;
// }